src/listfunc.c

```diff
@@ -1460,6 +1463,121 @@ static Obj FuncLIST_WITH_IDENTICAL_ENTRIES(Obj self, Obj n, Obj obj)
     return list;
 }
 
+static Obj FastCallFuncList(Obj func, Obj list)
+{
+    switch (LEN_PLIST(list)) {
+    case 0:
+        return CALL_0ARGS(func);
+    case 1:
+        return CALL_1ARGS(func, ELM_PLIST(list, 1));
+    case 2:
+        return CALL_2ARGS(func, ELM_PLIST(list, 1), ELM_PLIST(list, 2));
+    case 3:
+        return CALL_3ARGS(func, ELM_PLIST(list, 1), ELM_PLIST(list, 2),
+                          ELM_PLIST(list, 3));
+    case 4:
+        return CALL_4ARGS(func, ELM_PLIST(list, 1), ELM_PLIST(list, 2),
+                          ELM_PLIST(list, 3), ELM_PLIST(list, 4));
+    case 5:
+        return CALL_5ARGS(func, ELM_PLIST(list, 1), ELM_PLIST(list, 2),
+                          ELM_PLIST(list, 3), ELM_PLIST(list, 4),
+                          ELM_PLIST(list, 5));
+    case 6:
+        return CALL_6ARGS(func, ELM_PLIST(list, 1), ELM_PLIST(list, 2),
+                          ELM_PLIST(list, 3), ELM_PLIST(list, 4),
+                          ELM_PLIST(list, 5), ELM_PLIST(list, 6));
+    default:
+        return CALL_XARGS(func, list);
+    }
+}
+
+static Obj IsListOrCollection;
+
+// TODO: document this a bit
+static int FoldLeftXHelp(Obj   gens,
+                         Obj   foldFunc,
+                         Obj * acc,
+                         Obj   abortValue,
+                         Obj   args,
+                         int   genIndex,
+                         int   valIndex)
+{
+    while (genIndex <= LEN_PLIST(gens)) {
+        Obj gen = ELM_PLIST(gens, genIndex);
+        if (IS_FUNC(gen))
+            gen = FastCallFuncList(gen, args);
+        if (gen == True)
+            genIndex++;
+        else if (gen == False)
+            return 0;
+        else if (IS_LIST(gen)) {
+            const int len = LEN_LIST(gen);
+            for (int i = 1; i <= len; i++) {
+                Obj elm = ELM0_LIST(gen, i);
+                if (!elm)
+                    continue; // skip holes
+                AssPlist(args, valIndex, elm);
+                if (FoldLeftXHelp(gens, foldFunc, acc, abortValue, args,
+                                  genIndex + 1, valIndex + 1))
+                    return 1;
+            }
+            UNB_LIST(args, valIndex);
+            return 0;
+        }
+        else if (CALL_1ARGS(IsListOrCollection, gen) == True) {
```

Comment:
actually here is a chance to generalize `ListX` and friends to also accept iterators: we just need to accept objects in the filter `IsIterator` here, that should do it. Note that `List` and friends already (mostly?) support iterators.